# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/ipc/CanvasChild.cpp
# Commit: f1db1cd01b38
# Full Hash: f1db1cd01b380dcfc65ee07dcb851c4d70058175
# Author: Bob Owen <bobowencode@gmail.com>
# Date: 2023-09-07 04:09:51
# Regressor Bug: 1851636
# File Overlap Count: 1
# Description:
#   Bug 1851636: Don't create a new buffer if we fail to process the drop buffer event. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D187558
# ==============================================================================

diff -r 1db7a958b945 -r f1db1cd01b38 gfx/layers/ipc/CanvasChild.cpp
--- a/gfx/layers/ipc/CanvasChild.cpp	Wed Sep 06 16:37:55 2023 +0000
+++ b/gfx/layers/ipc/CanvasChild.cpp	Wed Sep 06 16:37:56 2023 +0000
@@ -231,10 +231,10 @@
   mRecorder->TakeExternalSurfaces(mLastTransactionExternalSurfaces);
 }
 
-void CanvasChild::EnsureBeginTransaction() {
+bool CanvasChild::EnsureBeginTransaction() {
   // We drop mRecorder in ActorDestroy to break the reference cycle.
   if (!mRecorder) {
-    return;
+    return false;
   }
 
   if (!mIsInTransaction) {
@@ -245,13 +245,15 @@
       if (!mRecorder->SwitchBuffer(OtherPid(), &handle) ||
           !SendNewBuffer(std::move(handle))) {
         mRecorder = nullptr;
-        return;
+        return false;
       }
     }
 
     mRecorder->RecordEvent(RecordedCanvasBeginTransaction());
     mIsInTransaction = true;
   }
+
+  return true;
 }
 
 void CanvasChild::EndTransaction() {
@@ -324,7 +326,11 @@
   if (!mIsInTransaction) {
     mTransactionsSinceGetDataSurface = 0;
   }
-  EnsureBeginTransaction();
+
+  if (!EnsureBeginTransaction()) {
+    return nullptr;
+  }
+
   mRecorder->RecordEvent(RecordedPrepareDataForSurface(aSurface));
   uint32_t checkpoint = mRecorder->CreateCheckpoint();
 